File: AIAgent/ZBuild/Repos/cxdb-main/clients/go/fs.go
// Copyright 2025 StrongDM Inc
// SPDX-License-Identifier: Apache-2.0

package cxdb

import (
	"bytes"
	"context"
	"encoding/binary"
	"fmt"
	"time"

	"github.com/zeebo/blake3"
)

// Protocol message types for filesystem operations
const (
	msgAttachFs uint16 = 10
	msgPutBlob  uint16 = 11
)

// AttachFsRequest contains parameters for attaching a filesystem snapshot to a turn.
type AttachFsRequest struct {
	// TurnID is the turn to attach the snapshot to.
	TurnID uint64

	// FsRootHash is the BLAKE3-256 hash of the root tree object.
	FsRootHash [32]byte
}

// AttachFsResult contains the result of an attach operation.
type AttachFsResult struct {
	TurnID     uint64
	FsRootHash [32]byte
}

// AttachFs attaches a filesystem snapshot to an existing turn.
// The tree objects and file blobs must already exist in the blob store.
func (c *Client) AttachFs(ctx context.Context, req *AttachFsRequest) (*AttachFsResult, error) {
	payload := &bytes.Buffer{}
	_ = binary.Write(payload, binary.LittleEndian, req.TurnID)
	payload.Write(req.FsRootHash[:])

	resp, err := c.sendRequest(ctx, msgAttachFs, payload.Bytes())
	if err != nil {
		return nil, fmt.Errorf("attach fs: %w", err)
	}

	if len(resp.payload) < 40 {
		return nil, fmt.Errorf("%w: attach fs response too short (%d bytes)", ErrInvalidResponse, len(resp.payload))
	}

	result := &AttachFsResult{
		TurnID: binary.LittleEndian.Uint64(resp.payload[0:8]),
	}
	copy(result.FsRootHash[:], resp.payload[8:40])

	return result, nil
}

// PutBlobRequest contains parameters for storing a blob.
type PutBlobRequest struct {
	// Data is the raw blob content.
	Data []byte
}

// PutBlobResult contains the result of a put blob operation.
type PutBlobResult struct {
	// Hash is the BLAKE3-256 hash of the blob.
	Hash [32]byte

	// WasNew indicates whether this was a new blob (true) or already existed (false).
	WasNew bool
}

// PutBlob stores a blob in the content-addressed store.
// The hash is computed from the data and verified by the server.
func (c *Client) PutBlob(ctx context.Context, req *PutBlobRequest) (*PutBlobResult, error) {
	// Compute hash
	hash := blake3.Sum256(req.Data)

	payload := &bytes.Buffer{}
	payload.Write(hash[:])
	_ = binary.Write(payload, binary.LittleEndian, uint32(len(req.Data)))
	payload.Write(req.Data)

	resp, err := c.sendRequest(ctx, msgPutBlob, payload.Bytes())
	if err != nil {
		return nil, fmt.Errorf("put blob: %w", err)
	}

	if len(resp.payload) < 33 {
		return nil, fmt.Errorf("%w: put blob response too short (%d bytes)", ErrInvalidResponse, len(resp.payload))
	}

	result := &PutBlobResult{
		WasNew: resp.payload[32] == 1,
	}
	copy(result.Hash[:], resp.payload[0:32])

	return result, nil
}

// PutBlobIfAbsent stores a blob only if it doesn't already exist.
// Returns the hash and whether the blob was stored.
func (c *Client) PutBlobIfAbsent(ctx context.Context, data []byte) ([32]byte, bool, error) {
	result, err := c.PutBlob(ctx, &PutBlobRequest{Data: data})
	if err != nil {
		return [32]byte{}, false, err
	}
	return result.Hash, result.WasNew, nil
}

// AppendTurnWithFs appends a new turn with an optional filesystem snapshot.
// If fsRootHash is non-nil, the filesystem snapshot will be attached to the turn.
func (c *Client) AppendTurnWithFs(ctx context.Context, req *AppendRequest, fsRootHash *[32]byte) (*AppendResult, error) {
	encoding := req.Encoding
	if encoding == 0 {
		encoding = EncodingMsgpack
	}
	compression := req.Compression

	// Compute BLAKE3 hash of payload
	hash := blake3.Sum256(req.Payload)

	payload := &bytes.Buffer{}
	_ = binary.Write(payload, binary.LittleEndian, req.ContextID)
	_ = binary.Write(payload, binary.LittleEndian, req.ParentTurnID)

	_ = binary.Write(payload, binary.LittleEndian, uint32(len(req.TypeID)))
	payload.WriteString(req.TypeID)
	_ = binary.Write(payload, binary.LittleEndian, req.TypeVersion)

	_ = binary.Write(payload, binary.LittleEndian, encoding)
	_ = binary.Write(payload, binary.LittleEndian, compression)
	_ = binary.Write(payload, binary.LittleEndian, uint32(len(req.Payload))) // uncompressed len
	payload.Write(hash[:])

	_ = binary.Write(payload, binary.LittleEndian, uint32(len(req.Payload)))
	payload.Write(req.Payload)

	_ = binary.Write(payload, binary.LittleEndian, uint32(len(req.IdempotencyKey)))
	if len(req.IdempotencyKey) > 0 {
		payload.WriteString(req.IdempotencyKey)
	}

	// If fsRootHash is provided, append it and set flags
	var flags uint16
	if fsRootHash != nil {
		flags = 1 // bit 0 = has_fs_root
		payload.Write(fsRootHash[:])
	}

	resp, err := c.sendRequestWithFlags(ctx, msgAppend, flags, payload.Bytes())
	if err != nil {
		return nil, fmt.Errorf("append turn: %w", err)
	}

	if len(resp.payload) < 52 {
		return nil, fmt.Errorf("%w: append response too short (%d bytes)", ErrInvalidResponse, len(resp.payload))
	}

	result := &AppendResult{
		ContextID: binary.LittleEndian.Uint64(resp.payload[0:8]),
		TurnID:    binary.LittleEndian.Uint64(resp.payload[8:16]),
		Depth:     binary.LittleEndian.Uint32(resp.payload[16:20]),
	}
	copy(result.PayloadHash[:], resp.payload[20:52])

	return result, nil
}

// sendRequestWithFlags is like sendRequest but allows setting custom flags.
func (c *Client) sendRequestWithFlags(ctx context.Context, msgType uint16, flags uint16, payload []byte) (*frame, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.closed {
		return nil, ErrClientClosed
	}

	// Set deadline for this request
	deadline := time.Now().Add(c.timeout)
	if d, ok := ctx.Deadline(); ok && d.Before(deadline) {
		deadline = d
	}
	if err := c.conn.SetDeadline(deadline); err != nil {
		return nil, fmt.Errorf("set deadline: %w", err)
	}
	defer func() { _ = c.conn.SetDeadline(time.Time{}) }() // Clear deadline

	reqID := c.reqID.Add(1)

	if err := c.writeFrameWithFlags(msgType, flags, reqID, payload); err != nil {
		return nil, err
	}

	resp, err := c.readFrame()
	if err != nil {
		return nil, err
	}

	if resp.msgType == msgError {
		return nil, parseServerError(resp.payload)
	}

	return resp, nil
}

func (c *Client) writeFrameWithFlags(msgType uint16, flags uint16, reqID uint64, payload []byte) error {
	header := &bytes.Buffer{}
	_ = binary.Write(header, binary.LittleEndian, uint32(len(payload)))
	_ = binary.Write(header, binary.LittleEndian, msgType)
	_ = binary.Write(header, binary.LittleEndian, flags)
	_ = binary.Write(header, binary.LittleEndian, reqID)

	_, err := c.conn.Write(append(header.Bytes(), payload...))
	return err
}

File: AIAgent/ZBuild/Repos/cxdb-main/clients/rust/src/fs.rs
// Copyright 2025 StrongDM Inc
// SPDX-License-Identifier: Apache-2.0

use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
use std::io::Read;

use crate::client::{Client, RequestContext};
use crate::error::{Error, Result};
use crate::protocol::{ENCODING_MSGPACK, MSG_APPEND_TURN, MSG_ATTACH_FS, MSG_PUT_BLOB};
use crate::turn::{AppendRequest, AppendResult};

#[derive(Debug, Clone)]
pub struct AttachFsRequest {
    pub turn_id: u64,
    pub fs_root_hash: [u8; 32],
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AttachFsResult {
    pub turn_id: u64,
    pub fs_root_hash: [u8; 32],
}

#[derive(Debug, Clone)]
pub struct PutBlobRequest {
    pub data: Vec<u8>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PutBlobResult {
    pub hash: [u8; 32],
    pub was_new: bool,
}

impl Client {
    pub fn attach_fs(&self, ctx: &RequestContext, req: &AttachFsRequest) -> Result<AttachFsResult> {
        let mut payload = Vec::with_capacity(40);
        payload.write_u64::<LittleEndian>(req.turn_id)?;
        payload.extend_from_slice(&req.fs_root_hash);

        let frame = self.send_request(ctx, MSG_ATTACH_FS, &payload)?;
        if frame.payload.len() < 40 {
            return Err(Error::invalid_response(format!(
                "attach fs response too short ({} bytes)",
                frame.payload.len()
            )));
        }

        let mut cursor = std::io::Cursor::new(frame.payload);
        let turn_id = cursor.read_u64::<LittleEndian>()?;
        let mut hash = [0u8; 32];
        cursor.read_exact(&mut hash)?;

        Ok(AttachFsResult {
            turn_id,
            fs_root_hash: hash,
        })
    }

    pub fn put_blob(&self, ctx: &RequestContext, req: &PutBlobRequest) -> Result<PutBlobResult> {
        let hash = blake3::hash(&req.data);
        let mut payload = Vec::with_capacity(36 + req.data.len());
        payload.extend_from_slice(hash.as_bytes());
        payload.write_u32::<LittleEndian>(req.data.len() as u32)?;
        payload.extend_from_slice(&req.data);

        let frame = self.send_request(ctx, MSG_PUT_BLOB, &payload)?;
        if frame.payload.len() < 33 {
            return Err(Error::invalid_response(format!(
                "put blob response too short ({} bytes)",
                frame.payload.len()
            )));
        }
        let mut hash_bytes = [0u8; 32];
        hash_bytes.copy_from_slice(&frame.payload[0..32]);
        let was_new = frame.payload[32] == 1;
        Ok(PutBlobResult {
            hash: hash_bytes,
            was_new,
        })
    }

    pub fn put_blob_if_absent(
        &self,
        ctx: &RequestContext,
        data: Vec<u8>,
    ) -> Result<([u8; 32], bool)> {
        let result = self.put_blob(ctx, &PutBlobRequest { data })?;
        Ok((result.hash, result.was_new))
    }

    pub fn append_turn_with_fs(
        &self,
        ctx: &RequestContext,
        req: &AppendRequest,
        fs_root_hash: Option<[u8; 32]>,
    ) -> Result<AppendResult> {
        let encoding = if req.encoding == 0 {
            ENCODING_MSGPACK
        } else {
            req.encoding
        };

        let hash = blake3::hash(&req.payload);
        let mut payload = Vec::with_capacity(128 + req.payload.len());
        payload.write_u64::<LittleEndian>(req.context_id)?;
        payload.write_u64::<LittleEndian>(req.parent_turn_id)?;
        payload.write_u32::<LittleEndian>(req.type_id.len() as u32)?;
        payload.extend_from_slice(req.type_id.as_bytes());
        payload.write_u32::<LittleEndian>(req.type_version)?;
        payload.write_u32::<LittleEndian>(encoding)?;
        payload.write_u32::<LittleEndian>(req.compression)?;
        payload.write_u32::<LittleEndian>(req.payload.len() as u32)?;
        payload.extend_from_slice(hash.as_bytes());
        payload.write_u32::<LittleEndian>(req.payload.len() as u32)?;
        payload.extend_from_slice(&req.payload);
        payload.write_u32::<LittleEndian>(req.idempotency_key.len() as u32)?;
        if !req.idempotency_key.is_empty() {
            payload.extend_from_slice(&req.idempotency_key);
        }

        let mut flags = 0u16;
        if let Some(hash) = fs_root_hash {
            flags |= 1;
            payload.extend_from_slice(&hash);
        }

        let frame = self.send_request_with_flags(ctx, MSG_APPEND_TURN, flags, &payload)?;
        if frame.payload.len() < 52 {
            return Err(Error::invalid_response(format!(
                "append response too short ({} bytes)",
                frame.payload.len()
            )));
        }
        let mut cursor = std::io::Cursor::new(frame.payload);
        let context_id = cursor.read_u64::<LittleEndian>()?;
        let turn_id = cursor.read_u64::<LittleEndian>()?;
        let depth = cursor.read_u32::<LittleEndian>()?;
        let mut hash = [0u8; 32];
        cursor.read_exact(&mut hash)?;
        Ok(AppendResult {
            context_id,
            turn_id,
            depth,
            payload_hash: hash,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_util::{decode_hex, load_fixture};

    fn build_append_payload(req: &AppendRequest, fs_root_hash: Option<[u8; 32]>) -> Vec<u8> {
        let encoding = if req.encoding == 0 {
            ENCODING_MSGPACK
        } else {
            req.encoding
        };
        let hash = blake3::hash(&req.payload);
        let mut payload = Vec::new();
        payload.write_u64::<LittleEndian>(req.context_id).unwrap();
        payload
            .write_u64::<LittleEndian>(req.parent_turn_id)
            .unwrap();
        payload
            .write_u32::<LittleEndian>(req.type_id.len() as u32)
            .unwrap();
        payload.extend_from_slice(req.type_id.as_bytes());
        payload.write_u32::<LittleEndian>(req.type_version).unwrap();
        payload.write_u32::<LittleEndian>(encoding).unwrap();
        payload.write_u32::<LittleEndian>(req.compression).unwrap();
        payload
            .write_u32::<LittleEndian>(req.payload.len() as u32)
            .unwrap();
        payload.extend_from_slice(hash.as_bytes());
        payload
            .write_u32::<LittleEndian>(req.payload.len() as u32)
            .unwrap();
        payload.extend_from_slice(&req.payload);
        payload
            .write_u32::<LittleEndian>(req.idempotency_key.len() as u32)
            .unwrap();
        if !req.idempotency_key.is_empty() {
            payload.extend_from_slice(&req.idempotency_key);
        }
        if let Some(hash) = fs_root_hash {
            payload.extend_from_slice(&hash);
        }
        payload
    }

    #[test]
    fn fs_payloads_match_fixtures() {
        let fixture = load_fixture("attach_fs");
        assert_eq!(fixture.msg_type, MSG_ATTACH_FS);
        assert_eq!(fixture.flags, 0);
        let mut payload = Vec::new();
        payload.write_u64::<LittleEndian>(99).unwrap();
        payload.extend_from_slice(&[0xAA; 32]);
        assert_eq!(decode_hex(&fixture.payload_hex), payload);

        let fixture = load_fixture("put_blob");
        assert_eq!(fixture.msg_type, MSG_PUT_BLOB);
        assert_eq!(fixture.flags, 0);
        let data = b"hello blob";
        let hash = blake3::hash(data);
        let mut payload = Vec::new();
        payload.extend_from_slice(hash.as_bytes());
        payload
            .write_u32::<LittleEndian>(data.len() as u32)
            .unwrap();
        payload.extend_from_slice(data);
        assert_eq!(decode_hex(&fixture.payload_hex), payload);

        let fixture = load_fixture("append_with_fs");
        assert_eq!(fixture.msg_type, MSG_APPEND_TURN);
        assert_eq!(fixture.flags, 1);
        let req = AppendRequest {
            context_id: 1,
            parent_turn_id: 0,
            type_id: "cxdb.ConversationItem".into(),
            type_version: 3,
            payload: vec![0x91, 0x04],
            idempotency_key: Vec::new(),
            encoding: ENCODING_MSGPACK,
            compression: 0,
        };
        let payload = build_append_payload(&req, Some([0xBB; 32]));
        assert_eq!(decode_hex(&fixture.payload_hex), payload);
    }
}
